src/interface.jl

```diff
@@ -2,83 +2,101 @@
 
 
 """
-    isdensitytype(type::Type)
+    hasdensity(d)::Bool
 
-Returns `true` if `type` is compatible with the `DensityInterface` interface,
-otherwise returns `false` by default.
+Return `true` if `d` is compatible with the `DensityInterface` interface.
+
+`hasdensity(d) == true` implies that `d` is either a density itself or has an
+associated density, e.g. a probability density function or a Radon–Nikodym
+derivative with respect to an implicit base measure. It also implies that the
+value of that density at given points can be calculated via
+[`logdensityof`](@ref) and [`densityof`](@ref).
+```
 """
-function isdensitytype end
-export isdensitytype
+function hasdensity end
+export hasdensity
 
-@inline isdensitytype(::Type) = false
+@inline hasdensity(::Any) = false
 
-function check_if_densitytype(type::Type)
-    isdensitytype(type) || throw("Type $(type) is not compatible with DensityInterface")
+function check_hasdensity(d)
+    hasdensity(d) || throw("Object of type $(typeof(d)) is not compatible with DensityInterface")
 end
 
 
 """
-    logdensityof(density, x)::Real
-    logdensityof(density)
+    logdensityof(d, x)::Real
 
-Computes the logarithmic value of `density` at a given point `x`, resp.
-returns a function that does so:
+Computes the logarithmic value of density `d` or it's associated density
+at a given point `x`.
 
-```julia
-logy = logdensityof(some_density, x)
-logdensityof(some_density, x) == logdensityof(some_density)(x)
-```
+```jldoctest a
+julia> hasdensity(d)
+true
 
-and
-
-```julia
-log_f = logdensityof(density)
-log_f(x) == logdensityof(density, x)
-logfuncdensity(log_f) == density
+julia> logy = logdensityof(d, x); logy isa Real
+true
 ```
 
-`logdensityof(density)` defaults to `Base.Fix1(logdensityof, density)`, but
-may be specialized for a density type. If so, [`logfuncdensity`](@ref) will
-typically have to be specialized for the return type of `logdensityof` as
-well.
-    
-The following identity must always hold:
-
-```julia
-logfuncdensity(logdensityof(density)) == density
-```    
+See also [`hasdensity`](@ref) and [`densityof`](@ref).
 """
 function logdensityof end
 export logdensityof
 
-function logdensityof(density)
-    check_if_densitytype(typeof(density))
-    Base.Fix1(logdensityof, density)
+"""
+    logdensityof(d)
+
+Return a function that computes the logarithmic value of density `d`
+or it's associated density at a given points.
```

Comment:
In my draft for LogDensity.jl, I had added this paragraph to the docstring: 

```markdown
`logdensity` should be well-defined over the extension of `μ`'s support to the whole range of the
type of compatible `x` values, returning an equivalent of `-Inf` outside the support (for meaningful
values, ignoring \"undefined\" cases like `NaN`)".
```

Perhaps that's a useful requirement to be part of the interface here, too?

Reply:
I did have this requirement in BAT.jl originally (the function had a different name, but the same semantics). I dropped it though, because in Bayesian applications the prior provides enough information for the algorithm to never go to places outside of the support of the likelihood. This lowers the burden on the user implementing the likelihood (density).

So there, the `-Inf` is implicit, not explicit. Not every application will be Bayesian, resp. have a prior, of course. But I feel requiring an explicit `-Inf` return value outside of the support is too restrictive.

Reply:
```suggestion
or its associated density at a given points.
```

Reply:
MeausreTheory requires that `logdensity` evaluates to `-Inf` outside the support. In Bayesian applications, it's true that this doesn't matter **if** you're using the Stan-like approach of mapping everything to an unbounded real vector space.

But not every inference method works like this. Rejection sampling, particle filtering, MH, and many more sometimes walk outside the support, needing a `-Inf` "push" to get them back into line.

On the other hand, testing for this does add some overhead. I think I'd suggest a trait-based approach for this. That way, you can have the safety when you need it, and get better performance when you can guarantee statically that you won't leave the support.

Reply:
> I think I'd suggest a trait-based approach for this.

I guess that would tie in with my ideas regarding traits to determine supports/domains of functions, likely in a separate package since that has broad applications? Maybe such an API could then provide a function wrapper `with_default_outside_support(f, default_value)` or so?

Applications that do require an explicit `-Inf` could then use constucts like `f_to_eval = with_default_outside_support(logdensityof(d), -Inf)`

Reply:
There are also cases where returning an explicit `-Inf` from `logdensityof` is very difficult to realize, technically. We recently did a lot of work with inference on models that include DiffEq runs. For non-physical parameters, the DiffEq-solver fails to converge and aborts - but there's no simple criterion regarding what non-physical parameter values are. So what's unphysical or not becomes part of our assumptions, and so goes in the prior (which prevents the model from being run on such values). But implementing it in the model would mean having part of the prior repeated inside the model/likelihood, kinda, that would be awkward. Or we'd have to use try/catch, which would be even more awkward (and incompatible with Zygote). :-)

Reply:
> I guess that would tie in with my ideas regarding traits to determine supports/domains of functions, likely in a separate package since that has broad applications? Maybe such an API could then provide a function wrapper `with_default_outside_support(f, default_value)` or so?
>
> Applications that do require an explicit `-Inf` could then use constucts like `f_to_eval = with_default_outside_support(logdensityof(d), -Inf)`

I guess this would also require a `support(f)` function, implying that `f` "knows" its support. Is that what you're thinking? Also, we may know that we're in the support of `f`, or we may know that `f` already provides a proper `-Inf`. In those cases, the support check is extra overhead. So I think the challenge here is to set up the dispatch in a way that optimizes all the cases.

BTW in MeasureTheory we have a `restrict` function that produces a `RestrictedMeasure`. We can use this as a base measure, which takes care of all the `-Inf` stuff. It's not yet fully optimized, but we have a good start on it. But I'd be open to refactoring if we can get to a common interface without sacrificing expressiveness or performance.

Reply:
I think in general having such independent "get-additional-information-in-respect-to-..." traits for functions, measure and the like is a good appoach. We should have more of them. :-)

> BTW in MeasureTheory we have a restrict function that produces a RestrictedMeasure

Nice!

I'll mark this resolved then (please feel free to unresolve if not).